providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,922 @@
+        return 0;
+    }
+
+    *siglen = ret;
+    return 1;
+}
+
+static int rsa_verify_recover(void *vprsactx,
+                              unsigned char *rout,
+                              size_t *routlen,
+                              size_t routsize,
+                              const unsigned char *sig,
+                              size_t siglen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (rout == NULL) {
+        *routlen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (prsactx->pad_mode == RSA_X931_PADDING) {
+            if (!setup_tbuf(prsactx))
+                return 0;
+            ret = RSA_public_decrypt(siglen, sig, prsactx->tbuf, prsactx->rsa,
+                                     RSA_X931_PADDING);
+            if (ret < 1) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret--;
+            if (prsactx->tbuf[ret] != RSA_X931_hash_id(EVP_MD_type(prsactx->md))) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_ALGORITHM_MISMATCH);
+                return 0;
+            }
+            if (ret != EVP_MD_size(prsactx->md)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+                return 0;
+            }
+            if (rout && (routsize == 0 || routsize >= (size_t)ret))
+                memcpy(rout, prsactx->tbuf, ret);
+        } else if (prsactx->pad_mode == RSA_PKCS1_PADDING) {
+            size_t sltmp;
+
+            ret = int_rsa_verify(EVP_MD_type(prsactx->md),
+                                 NULL, 0, rout, &sltmp,
+                                 sig, siglen, prsactx->rsa);
+            if (ret <= 0) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret = sltmp;
+        } else {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE);
+            return 0;
+        }
+    } else {
+        ret = RSA_public_decrypt(siglen, sig, rout, prsactx->rsa,
+                                 prsactx->pad_mode);
+    }
+    if (ret < 0) {
+        ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+        return 0;
+    }
+    *routlen = ret;
+    return 1;
+}
+
+static int rsa_verify(void *vprsactx, const unsigned char *sig, size_t siglen,
+                      const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    size_t rslen;
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (prsactx->pad_mode == RSA_PKCS1_PADDING) {
+            if (!RSA_verify(EVP_MD_type(prsactx->md), tbs, tbslen,
+                            sig, siglen, prsactx->rsa)) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            return 1;
+        }
+
+        if (tbslen != prsactx->mdsize) {
```

Comment:
Shouldn't this check happen before we call RSA_verify?

Reply:
No.  `RSA_verify()` has its own verification, and has more complex checks of this length.

Reply:
In which case is this more of an assertion?

Reply:
Again, I blame rsa_pmeth.c:

https://github.com/openssl/openssl/blob/d3a27c5ee45a29edd9c0d60ad5929f67996f89fd/crypto/rsa/rsa_pmeth.c#L251-L277

Do note, though, that in that piece of code, `RSA_verify_PKCS1_PSS_mgf1()` is called with `tbs`, but `tbslen` isn't passed (gah!  Why the hell do we have such inconsistent APIs?????), so I suppose that's what this check protects.

Reply:
So maybe I should move the check to make it more obvious what it protects

Reply:
Done